driver-core/src/main/com/mongodb/internal/authentication/AwsInternalCredential.java

```diff
@@ -0,0 +1,211 @@
+
+public final class AwsInternalCredential {
+    private final String accessKeyId;
+    private final String secretAccessKey;
+    private final String sessionToken;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param accessKeyId     the non-null access key ID that identifies the temporary security credentials.
+     * @param secretAccessKey the non-null secret access key that can be used to sign requests
+     * @param sessionToken    the non-null session token
+     */
+    public AwsInternalCredential(@Nullable final String accessKeyId, @Nullable final String secretAccessKey,
+            @Nullable final String sessionToken) {
+        this.accessKeyId = accessKeyId;
+        this.secretAccessKey = secretAccessKey;
+        this.sessionToken = sessionToken;
+    }
+
+    /**
+     * Gets the access key ID that identifies the temporary security credentials.
+     *
+     * @return the accessKeyId, which may not be null
+     */
+    @Nullable
+    public String getAccessKeyId() {
+        return accessKeyId;
+    }
+
+    /**
+     * Gets the secret access key that can be used to sign requests.
+     *
+     * @return the secretAccessKey, which may not be null
+     */
+    @Nullable
+    public String getSecretAccessKey() {
+        return secretAccessKey;
+    }
+
+    /**
+     * Gets the session token.
+     *
+     * @return the sessionToken, which may not be null
+     */
+    @Nullable
+    public String getSessionToken() {
+        return sessionToken;
+    }
+
+    public static AwsInternalCredential obtainFromEnvironment(final AwsInternalCredential startingCredential,
+            final Supplier<AwsCredential> awsCredentialSupplier) {
+
+        if (awsCredentialSupplier != null) {
+            return obtainFromSupplier(startingCredential, awsCredentialSupplier);
+        }
+
+        BsonDocument ec2OrEcsResponse = null;
+
+        String accessKeyId;
+        if (startingCredential.getAccessKeyId() != null) {
+            accessKeyId = startingCredential.getAccessKeyId();
+        } else {
+            accessKeyId = System.getenv("AWS_ACCESS_KEY_ID");
+            if (accessKeyId == null) {
+                ec2OrEcsResponse = getEc2OrEcsResponse();
+                accessKeyId = ec2OrEcsResponse.getString("AccessKeyId").getValue();
+            }
+        }
+
+        String secretAccessKey;
+        if (startingCredential.getSecretAccessKey() != null) {
+            secretAccessKey = startingCredential.getSecretAccessKey();
+        } else {
+            if (System.getenv("AWS_SECRET_ACCESS_KEY") != null) {
+                secretAccessKey = System.getenv("AWS_SECRET_ACCESS_KEY");
+            } else {
+                ec2OrEcsResponse = ec2OrEcsResponse == null ? getEc2OrEcsResponse() : ec2OrEcsResponse;
+                secretAccessKey = ec2OrEcsResponse.getString("SecretAccessKey").getValue();
+            }
+        }
+
+        String sessionToken;
+        if (startingCredential.getAccessKeyId() != null) {
```

Comment:
Do you mean to call `getAccessKeyId()` here? I think it is right but I am surprised.

Reply:
No, I think I meant this to be `getSessionToken`.

Reply:
I'm wrong.  I transcribed this correctly from the original.  It's equivalent to this code from `AwsAuthenticator#getSessionToken`:

```
            String token = credential.getMechanismProperty(AWS_SESSION_TOKEN_KEY, null);
            if (credential.getUserName() != null) {
                return token;
            }
            if (token != null) {
                throw new IllegalArgumentException("The connection string contains a session token without credentials");
            }
```

Reply:
The intention as I understand it is that a session token isn't allowed without an accessKeyId.  Is that correct?

Reply:
Yeah, session token is not allowed without a accessKeyId.

AWS Credentials are either
(accessKeyId, secretAccessKey) - aka regular, non expiring, must be allocated in AWS console, per user
OR
(accessKeyId, secretAccessKey, sessionToken) - aka temporary, expiring, retrieved from ECS/EKS/EC2/assume role

Any single element is no enough to authenticate. Also combinations like (secretAccessKey, sessionToken) and (accessKeyId, sessionToken) simply do not work.